castle/utils.py

```diff
@@ -1,20 +1,31 @@
 import copy
+from datetime import datetime
 
 def clone(dict_object):
     return copy.deepcopy(dict_object)
 
 def deep_merge(base, extra):
-    """Deeply two dictionaries, overriding existing keys in the base.
+    """
+    Deeply merge two dictionaries, overriding existing keys in the base.
+
     :param base: The base dictionary which will be merged into.
     :param extra: The dictionary to merge into the base. Keys from this
         dictionary will take precedence.
     """
-    for key in extra:
+    if extra is None:
+        return
+
+    for key, value in extra.items():
         # If the key represents a dict on both given dicts, merge the sub-dicts
-        if key in base and isinstance(base[key], dict)\
-                and isinstance(extra[key], dict):
-            deep_merge(base[key], extra[key])
-            continue
+        if value is None:
+            if key in base:
+                del base[key]
+        elif isinstance(base.get(key), dict) and isinstance(value, dict):
+            deep_merge(base[key], value)
+        else:
+            # Otherwise, set the key on the base to be the value of the extra.
+            base[key] = value
 
-        # Otherwise, set the key on the base to be the value of the extra.
-        base[key] = extra[key]
+def timestamp():
+    """Return an ISO8601 timestamp representing the current datetime in UTC."""
+    return datetime.utcnow().isoformat()[:-3]
```

Comment:
why truncate the microseconds?

also, on the off chance that the timestamp is generated where microseconds=0, `isoformat` actually omits them entirely. In that case, your slice will actually lop off the seconds.

```python
>>> from datetime import datetime
>>> d = datetime.utcnow().replace(microsecond=0)
>>> d.isoformat()
'2018-02-08T17:34:53'
>>> d.isoformat()[:-3]
'2018-02-08T17:34'
```

Reply:
Separately, the timestamp lacks the timezone offset. While we know we requested the timestamp in UTC, sending it over the wire this way is ambiguous and feels like the wrong thing to do. The simplest thing to do would be to tack on  a `+00:00` to the end of the string.

Reply:
we are not removing microseconds `replace(microsecond=0)` in that method so they will be properly truncated and delivered to the api. But I totally agree with you. we should send full iso format no matter how it looks like, to make it more clear and readable and api will take care of proper formatting